### PR TITLE
add SupplierWrapper to support java 1.8+ CompletableFuture.supplyAsync #3779

### PR DESCRIPTION
**Pull request by sxzaihua**
Please answer these questions before submitting pull request

- Why submit this pull request?
- [3779 ] New feature provided

- Related issues
I want use async traceid like this:
CompletableFuture.supplyAsync(SupplierWrapper.of(()->getString1("a"))).thenAccept(System.out::println);

But now skywalking only supports callable(CallableWrapper), not supports supplier

### CODE REVIEW

**Comment by sxzaihua** [comment]
add SupplierWrapper to support java 1.8+ CompletableFuture.supplyAsync #3776

**Comment by wu-sheng** [comment]
Fix CI, and add integration plugin test for this new and callable/runnable too. The document will be ready today.

**Comment by arugal** [comment]
Hi, use case tests can be added here [apm-toolkit-trace-scenario](https://github.com/apache/skywalking/tree/master/test/plugin/scenarios/apm-toolkit-trace-scenario)  @sxzaihua

**Comment by wu-sheng** [comment]
> Hi, use case tests can be added here apm-toolkit-trace-scenario

This case didn't include the old Runnable and Callable from last time I checked. Those should be added too.

File: test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/toolkit/controller/TestController.java [hunk 1]
```diff
@@ -62,6 +62,14 @@ public String toolKitCase() {
                 // ignore
             }
         });
+        testService.asyncSupplier(()->{
+        	try {
+				visit("http://localhost:8080/apm-toolkit-trace-scenario/case/asyncVisit/runnable");
+			} catch (IOException e) {
```
Comment (arugal) [inline test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/toolkit/controller/TestController.java]: It should be ```http://localhost:8080/apm-toolkit-trace-scenario/case/asyncVisit/supplier ```

**Comment by arugal** [comment]
Hi, [Expecteddata.yaml](https://github.com/apache/skywalking/blob/master/test/plugin/scenarios/apm-toolkit-trace-scenario/config/expectedData.yaml) is not updated. @sxzaihua

**Comment by arugal** [comment]
> This case didn't include the old Runnable and Callable from last time I checked. Those should be added too.

[CallableWrapper](https://github.com/apache/skywalking/blob/master/test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/org/apache/skywalking/apm/toolkit/trace/CallableWrapper.java) and [RunnableWrapper](https://github.com/apache/skywalking/blob/master/test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/org/apache/skywalking/apm/toolkit/trace/RunnableWrapper.java) ?

**Comment by sxzaihua** [comment]
可以加个QQ聊下吗，381124590


| |
sxzaihua
|
|
sxzaihua@163.com
|
签名由网易邮箱大师定制
On 11/5/2019 13:37，zhangwei<notifications@github.com> wrote：

This case didn't include the old Runnable and Callable from last time I checked. Those should be added too.

CallableWrapper and RunnableWrapper ?

**Comment by wu-sheng** [comment]
The plugin test framework doc is ready, #3786. Please take a look

File: test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/toolkit/controller/TestController.java (hunk 1 above)
Comment (wu-sheng) [reply test/plugin/scenarios/apm-toolkit-trace-scenario/src/main/java/test/org/apache/skywalking/apm/testcase/toolkit/controller/TestController.java]: @sxzaihua You should add your own servlet method(`supplier`), and follow the plugin test doc, to add your new segment to verify the plugin feature.

**Comment by wu-sheng** [comment]
The test fails, please check locally.

**Comment by wu-sheng** [comment]
@arugal He changed this test, I don't know whether should be rerun.

**Comment by arugal** [comment]
> @arugal He changed this test, I don't know whether should be rerun.

```
[2019-11-05T08:53:23.451Z] start container of testcase.name=apm-toolkit-trace-scenario-1.0.0
[2019-11-05T08:53:23.451Z] Unable to find image 'skywalking/agent-test-jvm:wl2_559' locally
[2019-11-05T08:53:24.037Z] docker: Error response from daemon: pull access denied for skywalking/agent-test-jvm, repository does not exist or may require 'docker login': denied: requested access to the resource is denied.
[2019-11-05T08:53:24.037Z] See 'docker run --help'.
[2019-11-05T08:53:24.037Z] docker startup failure!
```
This is the cause of failure.

**Comment by wu-sheng** [comment]
One reminder for @sxzaihua, the document has not been updated.

**Comment by arugal** [comment]
LGTM

**Comment by wu-sheng** [comment]
Please update the doc.

**Comment by wu-sheng** [comment]
@arugal @kezhenxu94 @dmsolr Please recheck.

**Comment by arugal** [comment]
LGTM

**Pull request merged**